README.md
THUMDER
=======

<center>
  <img src="./assets/favicon.256x256.png" alt="ICON" width="10%" height="10%"/>
</center>

<center>
  <a href="https://angular.io/">
    <img src="https://www.vectorlogo.zone/logos/angular/angular-icon.svg" alt="Angular Logo" width="10%" height="10%"/>
  </a>
  <a href="https://electronjs.org/">
    <img src="https://www.vectorlogo.zone/logos/electronjs/electronjs-icon.svg" alt="Electron Logo">
  </a>
</center>

<center>
  <a href="https://app.netlify.com/sites/thumder/deploys">
    <img src="https://api.netlify.com/api/v1/badges/491cdbc6-34ea-4ef0-92c0-be4f2a2ec3cb/deploy-status" alt="Netlify Status">
  </a>
</center>

<center>
  Página del proyecto en producción <a href="https://thumder.netlify.com">THUMDER</a><br>
</center>

<center>
  Repositorio del servidor del proyecto <a href="https://github.com/nonodev96/THUMDER-server">THUMDER server</a>
</center>

## Instalación de las dependencias

```bash
sudo npm install -g  @angular/cli@12.5.0  # Instala la herramienta de desarrollo
npm install                               # Instala las dependencias
```

Tu necesitas aumenta la memoria de node, para ello se debe modificar la variable de entorno de node (`NODE_OPTIONS`),
para ello debemos asignar al menos 4Gb de memoria `--max_old_space_size=<size>`.

## Desarrollo


| **Commandos**                  | **Descripción**                                         |
|:-------------------------------|:--------------------------------------------------------|
| `npm run ng:serve-angular:dev` | Enciende el servidor de angular en modo de desarrollo   |
| `npm run ws:server`            | Enciende el servidor websocket, requiere THUMDER Server |

## Despliegue


| **Commandos**                         | **Descripción**                                                                                   |
|:--------------------------------------|:--------------------------------------------------------------------------------------------------|
| `npm run ng:build-angular:production` | Construye la aplicación angular de navegador con las variables de producción                      |
| `npm run electron:build`              | Construye la aplicación angular y encapsula la aplicación dentro electron generando un ejecutable |

El despliegue genera las carpetas `/dist`, `/dist-angular` y `/release`.

La carpeta `/dist-angular` es la indicada para el despliegue en servidores.

La carpeta `/release` contiene los ficheros ejecutables con la aplicación.

## Otros comandos


| **Commandos**                         | **Descripción** |
|:--------------------------------------|:----------------|
| `npm run ng:build:dev`                |                 |
| `npm run ng:build:web`                |                 |
| `npm run ng:build:production`         |                 |
| `npm run ng:build-angular:dev`        |                 |
| `npm run ng:build-angular:web`        |                 |
| `npm run ng:build-angular:production` |                 |
| `npm run ng:serve:dev`                |                 |
| `npm run ng:serve:web`                |                 |
| `npm run ng:serve:production`         |                 |
| `npm run ng:serve-angular:dev`        |                 |
| `npm run ng:serve-angular:web`        |                 |
| `npm run ng:serve-angular:production` |                 |

## Server e2e


| **Commandos**          | **Descripción**                                                                 |
|:-----------------------|:--------------------------------------------------------------------------------|
| `npm run cypress:open` | Abre la aplicación con cypress app y configura los comandos y pruebas           |
| `npm run cypress:run`  | Ejecuta las pruebas sobre el servidor (solo son visibles por linea de comandos) |

### Cobertura de navegadores

```bash
npx browserslist
npx browserslist --coverage
```

These browsers account for 86.79% of all users globally

## Versión actual de las dependencias

- Angular v12.5.0
- Electron v16.2.6
- Electron Builder v23.0.3


CHANGELOG.md
## <small>1.6.3 (2022-07-02)</small>

* Update readme, wiki and fix code request test ([5dd294a](https://github.com/nonodev96/THUMDER/commit/5dd294a))



## <small>1.6.2 (2022-06-28)</small>

* Fix GitHub Actions with firebase ([7f6699f](https://github.com/nonodev96/THUMDER/commit/7f6699f))



## <small>1.6.1 (2022-06-28)</small>

* Fix electron BrowserWindow and modify calculator view ([3d3b639](https://github.com/nonodev96/THUMDER/commit/3d3b639))



## 1.6.0 (2022-06-28)

* Fix with drag and drop, new component card with docs and names in UML sequences changes ([a11d21a](https://github.com/nonodev96/THUMDER/commit/a11d21a))
* Lazy loading with angular routing ([9208302](https://github.com/nonodev96/THUMDER/commit/9208302))
* Minor refactor in MachineService, ElectronService, PixiTHUMDER_CycleClockDiagram and PixiTHUMDER_Pip ([03534a5](https://github.com/nonodev96/THUMDER/commit/03534a5))
* PlantUML Gantt with all project with svg and pdf generator. ([df914a1](https://github.com/nonodev96/THUMDER/commit/df914a1))



## <small>1.3.9 (2022-03-31)</small>

* fix files in file system storage (delete item) ([86b65fa](https://github.com/nonodev96/THUMDER/commit/86b65fa))
* Refactor in all firebase method with new modular functions and providers for electron. ([f8de3f8](https://github.com/nonodev96/THUMDER/commit/f8de3f8))
* update wiki and multiview drag and drop ([6d42941](https://github.com/nonodev96/THUMDER/commit/6d42941))



## <small>1.3.8 (2022-03-09)</small>

* Add submodul wiki ([7d7a041](https://github.com/nonodev96/THUMDER/commit/7d7a041))
* CHANGELOG ([f04025e](https://github.com/nonodev96/THUMDER/commit/f04025e))
* Deleted submodule wiki ([41b6bb9](https://github.com/nonodev96/THUMDER/commit/41b6bb9))
* fix karma version ([ee905d3](https://github.com/nonodev96/THUMDER/commit/ee905d3))
* fix layout header ([a8d0230](https://github.com/nonodev96/THUMDER/commit/a8d0230))
* Wiki ([b737a0f](https://github.com/nonodev96/THUMDER/commit/b737a0f))
* Wiki assets 3 and color in cycle clock diagram ([66f67b5](https://github.com/nonodev96/THUMDER/commit/66f67b5))



## <small>1.3.6 (2022-02-14)</small>

* .browserslistrc, fix package.json, components, tests with karma and README ([d0c5fad](https://github.com/nonodev96/THUMDER/commit/d0c5fad))
* CHANGELOG ([c70702c](https://github.com/nonodev96/THUMDER/commit/c70702c))
* minor changes ([e0514eb](https://github.com/nonodev96/THUMDER/commit/e0514eb))
* polyfills for karma tests ([d6fa110](https://github.com/nonodev96/THUMDER/commit/d6fa110))



## <small>1.3.5 (2022-01-30)</small>

* Add submodule with wiki reference, ChartJS in statistics, fix filesystem, reset localstorage with lo ([312f833](https://github.com/nonodev96/THUMDER/commit/312f833))
* Add to angular.json glob Wiki and update github actions with recursive submodule ([4f3bf81](https://github.com/nonodev96/THUMDER/commit/4f3bf81))
* fixed Pixi arrow, pipeline, add favicon, add redirect to cookies, profile and save project ([79583c7](https://github.com/nonodev96/THUMDER/commit/79583c7))
* version 1.3.5 ([b2f9963](https://github.com/nonodev96/THUMDER/commit/b2f9963))



## <small>1.3.4 (2022-01-25)</small>

* Add icons, electron ipcMain, electron notification, toggle card with title, fix href and config cols ([80126c1](https://github.com/nonodev96/THUMDER/commit/80126c1))



## <small>1.3.3 (2022-01-24)</small>

* Refactor: clean code ([039c3f9](https://github.com/nonodev96/THUMDER/commit/039c3f9))
* FileSystem fixed and new Observables and performance, UML refactor, refactor in __core ([a9c0142](https://github.com/nonodev96/THUMDER/commit/a9c0142))



## <small>1.3.2 (2022-01-14)</small>

* add AppAdminLTE init, refactor in CONSTANTS, Types and routes, fix layout page, new layout in index, ([bce3545](https://github.com/nonodev96/THUMDER/commit/bce3545))
* fix config with default value ([c243f19](https://github.com/nonodev96/THUMDER/commit/c243f19))
* Multiview performance, Drag & Drop, card resize, items default in project, fix guard in path /, tran ([62e3ef2](https://github.com/nonodev96/THUMDER/commit/62e3ef2))



## <small>1.3.1 (2022-01-07)</small>

* Add Netlify Status and Translations, add breakpoints default value. ([f2442a7](https://github.com/nonodev96/THUMDER/commit/f2442a7))
* Update create-release.yaml ([d16f460](https://github.com/nonodev96/THUMDER/commit/d16f460))
* Update create-release.yaml ([f088500](https://github.com/nonodev96/THUMDER/commit/f088500))
* Update GitHub Action create release ([e86d5f9](https://github.com/nonodev96/THUMDER/commit/e86d5f9))
* Update GitHub Action create release v10 | node 14.18.2 | npm 6.14.15 | I hate .yaml in GitHub action ([36adc89](https://github.com/nonodev96/THUMDER/commit/36adc89))
* Update GitHub Action create release v11 | node 14.18.2 | npm 6.14.15 | I hate .yaml in GitHub action ([b2b60c1](https://github.com/nonodev96/THUMDER/commit/b2b60c1))
* Update GitHub Action create release v2 ([462500b](https://github.com/nonodev96/THUMDER/commit/462500b))
* Update GitHub Action create release v3 ([0ee6fed](https://github.com/nonodev96/THUMDER/commit/0ee6fed))
* Update GitHub Action create release v4 ([86ff143](https://github.com/nonodev96/THUMDER/commit/86ff143))
* Update GitHub Action create release v5 ([eeda95e](https://github.com/nonodev96/THUMDER/commit/eeda95e))
* Update GitHub Action create release v6 | node 14.18.2 | npm 6.14.15 ([8331ada](https://github.com/nonodev96/THUMDER/commit/8331ada))
* Update GitHub Action create release v7 | node 14.18.2 | npm 6.14.15 | I hate .yaml in GitHub actions ([1744fe0](https://github.com/nonodev96/THUMDER/commit/1744fe0))
* Update GitHub Action create release v8 | node 14.18.2 | npm 6.14.15 | I hate .yaml in GitHub actions ([ae18bee](https://github.com/nonodev96/THUMDER/commit/ae18bee))
* Update GitHub Action create release v9 | node 14.18.2 | npm 6.14.15 | I hate .yaml in GitHub actions ([2f4a827](https://github.com/nonodev96/THUMDER/commit/2f4a827))

src/app/Types.ts
import { InterfaceRegisters } from "./__core/DLX/interfaces";

declare global {
  interface Window {
    process: any;
    require: any;
    jQuery: any;
    $: any;
  }
}

export enum EnumLogLevel {
  All = 0,
  Debug = 1,
  Info = 2,
  Warn = 3,
  Error = 4,
  Fatal = 5,
  Off = 6
}

export enum EnumSeverity {
  Hint = 1,
  Info = 2,
  Warning = 4,
  Error = 8
}

export type TypeLogger = {
  index: number;
  value: any;
};

export type TypeComponentStatus = "OnInit" | "AfterViewInit" | "OnDestroy";

export type TypeOnKeyEvent = {
  domEvent: KeyboardEvent;
  key: string;
};

export type TypeExtrasIDE = {
  interfaceFileItem: InterfaceFileItem;
};

export interface InterfaceUser {
  uid: string;
  email: string;
  displayName: string;
  photoURL: string;
  emailVerified: boolean;
}

export interface InterfaceFileItem {
  $key: string; // ---? Firebase
  f_id: string; //-
  e1_uid: string; //-
  key: string; //-
  pathKeys: string[]; //-
  path: string;
  name: string;
  content: string;
  description: string;
  dateModified: Date;
  size: number;
  isDirectory: boolean;
  hasSubDirectories: boolean;
  thumbnail: string;
  dataItem: any;
}


export type StringOfLength<Min, Max> = string & {
  readonly StringOfLength: unique symbol // this is the phantom type
};

// This is a type guard function which can be used to assert that a string
// is of type StringOfLength<Min,Max>
function isStringOfLength<Min extends number, Max extends number>(str: string, min: Min, max: Max): str is StringOfLength<Min, Max> {
  return str.length >= min && str.length <= max;
}

// type constructor function
export function stringOfLength<Min extends number, Max extends number>(input: unknown, min: Min, max: Max): StringOfLength<Min, Max> {
  if (typeof input !== "string") {
    throw new Error("invalid input");
  }

  if (!isStringOfLength(input, min, max)) {
    throw new Error("input is not between specified min and max");
  }

  return input;
}

export type TypeTransformDecimalToBase = {
  base: number;
  maxLength?: number;
  fillString?: string
};

export type TypeTags = {
  line: number;
  content: string
}[];

export type TypeBreakpoints = {
  [line: number]: boolean
};

export type PublicRoutes_SET = {
  [name: string]: {
    lang: string;
    path: string;
    routerLink: string;
    displayName: string;
    show: boolean;
    data?: any;
    icon?: string;
    children?: PublicRoutes[];
  }
};

export type PublicRoutes = {
  lang: string;
  path: string;
  routerLink: string;
  displayName: string;
  data?: any;
  icon?: string;
  children?: PublicRoutes[];
};

export type TypeMemory = {
  address: string;
  value: number;
};

export type PublicRoutesList = PublicRoutes_SET;

export type TypeLine = number;

export type TypeAddress = `0x${string}`;

export type TypeAddressStage = {
  address: TypeAddress,
  stage: TypeStage
}

export type TypeData = "Byte" | "HalfWord" | "Word" | "Float" | "Double" | "ASCII";

export type TypeRegister = "Control" | "Integer" | "Float" | "Double";

export type TypePipelineStage = "IF" | "ID" | "intEX" | "MEM" | "WB" | "faddEX" | "fmultEX" | "fdivEX";

export type TypeStall = "Aborted" | "R-Stall" | "T-Stall" | "W-Stall" | "S-Stall" | "Stall";

export type TypeDataDisplayColumn = TypeData | "InstructionCode" | "Address-0-1-2-3" | "HalfWord-0-1"

export type TypeDataRepresentation = "Binary" | "Uint8Array";

export type TypeDirective =
  "GLOBAL"
  | "TEXT"
  | "SPACE"
  | "DATA"
  | "ALIGN"
  | "ASCII"
  | "ASCIIZ"
  | "BYTE"
  | "FLOAT"
  | "DOUBLE"
  | "WORD";

export type TypeIdTitleFile = {
  id: string,
  title: string,
  file: string
}

export type TypeDataRegister = {
  Control: {
    registers: number;
    size: number;
    maxLengthHexadecimal: number;
  };
  Integer: {
    registers: number;
    size: number;
    maxLengthHexadecimal: number;
  };
  Float: {
    registers: number;
    size: number;
    maxLengthHexadecimal: number;
  };
  Double: {
    registers: number;
    size: number;
    maxLengthHexadecimal: number;
  };
  DEFAULT: {
    registers: number;
    size: number;
    maxLengthHexadecimal: number;
  }
};

export type TypeRegisterControl =
  | "PC"
  | "IMAR"
  | "IR"
  | "A"
  | "AHI"
  | "B"
  | "BHI"
  | "BTA"
  | "ALU"
  | "ALUHI"
  | "FPSR"
  | "DMAR"
  | "SDR"
  | "SDRHI"
  | "LDR"
  | "LDRHI";

export type uint8 =
  | 0
  | 1
  | 2
  | 3
  | 4
  | 5
  | 6
  | 7
  | 8
  | 9
  | 10
  | 11
  | 12
  | 13
  | 14
  | 15
  | 16
  | 17
  | 18
  | 19
  | 20
  | 21
  | 22
  | 23
  | 24
  | 25
  | 26
  | 27
  | 28
  | 29
  | 30
  | 31
  | 32
  | 33
  | 34
  | 35
  | 36
  | 37
  | 38
  | 39
  | 40
  | 41
  | 42
  | 43
  | 44
  | 45
  | 46
  | 47
  | 48
  | 49
  | 50
  | 51
  | 52
  | 53
  | 54
  | 55
  | 56
  | 57
  | 58
  | 59
  | 60
  | 61
  | 62
  | 63
  | 64
  | 65
  | 66
  | 67
  | 68
  | 69
  | 70
  | 71
  | 72
  | 73
  | 74
  | 75
  | 76
  | 77
  | 78
  | 79
  | 80
  | 81
  | 82
  | 83
  | 84
  | 85
  | 86
  | 87
  | 88
  | 89
  | 90
  | 91
  | 92
  | 93
  | 94
  | 95
  | 96
  | 97
  | 98
  | 99
  | 100
  | 101
  | 102
  | 103
  | 104
  | 105
  | 106
  | 107
  | 108
  | 109
  | 110
  | 111
  | 112
  | 113
  | 114
  | 115
  | 116
  | 117
  | 118
  | 119
  | 120
  | 121
  | 122
  | 123
  | 124
  | 125
  | 126
  | 127
  | 128
  | 129
  | 130
  | 131
  | 132
  | 133
  | 134
  | 135
  | 136
  | 137
  | 138
  | 139
  | 140
  | 141
  | 142
  | 143
  | 144
  | 145
  | 146
  | 147
  | 148
  | 149
  | 150
  | 151
  | 152
  | 153
  | 154
  | 155
  | 156
  | 157
  | 158
  | 159
  | 160
  | 161
  | 162
  | 163
  | 164
  | 165
  | 166
  | 167
  | 168
  | 169
  | 170
  | 171
  | 172
  | 173
  | 174
  | 175
  | 176
  | 177
  | 178
  | 179
  | 180
  | 181
  | 182
  | 183
  | 184
  | 185
  | 186
  | 187
  | 188
  | 189
  | 190
  | 191
  | 192
  | 193
  | 194
  | 195
  | 196
  | 197
  | 198
  | 199
  | 200
  | 201
  | 202
  | 203
  | 204
  | 205
  | 206
  | 207
  | 208
  | 209
  | 210
  | 211
  | 212
  | 213
  | 214
  | 215
  | 216
  | 217
  | 218
  | 219
  | 220
  | 221
  | 222
  | 223
  | 224
  | 225
  | 226
  | 227
  | 228
  | 229
  | 230
  | 231
  | 232
  | 233
  | 234
  | 235
  | 236
  | 237
  | 238
  | 239
  | 240
  | 241
  | 242
  | 243
  | 244
  | 245
  | 246
  | 247
  | 248
  | 249
  | 250
  | 251
  | 252
  | 253
  | 254;

export type TypeMemoryIndexValue = {
  index: number;
  value: uint8;
};

export type TypeAllMemory = TypeMemoryIndexValue[];

export type TypeAllRegisters = {
  Control: {
    register: TypeRegisterControl;
    value: string; // 0x 00000000
  }[],
  Integer: {
    register: number;
    value: string; // 0x 00000000
  }[],
  Float: {
    register: number;
    value: string; // 0x 00000000
  }[],
};

export type TypeRegisterToEdit =
  "PC"
  | "IMAR"
  | "IR"
  | "A"
  | "AHI"
  | "B"
  | "BHI"
  | "BTA"
  | "ALU"
  | "ALUHI"
  | "FPSR"
  | "DMAR"
  | "SDR"
  | "SDRHI"
  | "LDR"
  | "LDRHI"

  | 0 | 10 | 20 | 30
  | 1 | 11 | 21 | 31
  | 2 | 12 | 22
  | 3 | 13 | 23
  | 4 | 14 | 24
  | 5 | 15 | 25
  | 6 | 16 | 26
  | 7 | 17 | 27
  | 8 | 18 | 28
  | 9 | 19 | 29;

export type TypeStage =
  ""
  | "IF"
  | "ID"
  | "intEX"
  | "MEM"
  | "WB"
  | "faddEX" | "fmultEX" | "fdivEX"
  | "trap"
  | "other"
  | "faddEX_0" | "fmultEX_0" | "fdivEX_0"
  | "faddEX_1" | "fmultEX_1" | "fdivEX_1"
  | "faddEX_2" | "fmultEX_2" | "fdivEX_2"
  | "faddEX_3" | "fmultEX_3" | "fdivEX_3"
  | "faddEX_4" | "fmultEX_4" | "fdivEX_4"
  | "faddEX_5" | "fmultEX_5" | "fdivEX_5"
  | "faddEX_6" | "fmultEX_6" | "fdivEX_6"
  | "faddEX_7" | "fmultEX_7" | "fdivEX_7";

export type TypeDirectiveData = {
  address: TypeAddress;     // 0x00000000
  hexValue: string;         // 0x00000000
  text: string;
  directive: TypeDirective;
};

export type TypeInstructionsData = {
  address: TypeAddress; // 0x00000000
  code: string;         // 0x00000000
  text: string;
  instruction: string;
};

export type TypeCodeResponse = {
  machineDirectives: TypeDirectiveData[],
  machineInstructions: TypeInstructionsData[],
};

export type TypeInstructionsData_Table = TypeInstructionsData & {
  row: number;
  stage?: TypeStage;
  index?: number;
};

export type TypeMultiviewConfiguration = {
  calculator: boolean;
  pipeline: boolean;
  cycle_clock_diagram: boolean;
  memory: boolean;
  registers: boolean;
  code: boolean;
  statistics: boolean;
  list_1: string[];
  list_2: string[];
};

export type TypeEnabledForwardingConfiguration = boolean;

export type TypeFloatingPointStageConfiguration = {
  addition: {
    count: number;
    delay: number;
  };
  multiplication: {
    count: number;
    delay: number;
  };
  division: {
    count: number;
    delay: number;
  }
};

export type TypeWebSocketConfiguration = {
  socket_url: string;
};

export type TypeLang = "en" | "sp";

export type TypeConfigurationMachine = {
  addition: {
    count: number;
    delay: number;
  },
  multiplication: {
    count: number;
    delay: number;
  },
  division: {
    count: number;
    delay: number;
  },
  memorySize: number;
  enabledForwarding: TypeEnabledForwardingConfiguration;
};

// DEBUG

export type TypeStatusMachine = {
  registers: InterfaceRegisters; // TODO Debug
  memory: TypeMemory[];
  breakpoints: TypeBreakpoints;
  [data: string]: any; // any to any
};

export type TypeRegisterToUpdate = {
  typeRegister: TypeRegister;
  register: string | TypeRegisterControl | TypeRegisterToEdit;
  hexadecimalValue: string;
};

export type TypeMemoryToUpdate = {
  address: TypeAddress;
  typeData: TypeData;
  value: string;
};

export type TypeCycleCell = {
  address: TypeAddress;
  addressRow: number;
  draw: boolean | TypeStall;
};

export type TypeCycleCellUnit = TypeCycleCell & {
  unit?: number
};

export type TypePipelineToProcess = TypeCycleCellUnit & {
  stage?: TypeStage;
};

export type TypeInstructionPipelineRepresentation = {
  text: string;
  draw: boolean | TypeStall;
}

export type TypeInstructionPipelineFloatingRepresentation = TypeInstructionPipelineRepresentation & {
  unit: number
}

export type TypeArrowCycle = {
  fromAddressRow: number;
  fromStep: number;
  toAddressRow: number;
  toStep: number;
  color: number; // hexadecimal 0xff00ff rgb
};

export type TypePipeline = {
  IF: TypeCycleCell;
  ID: TypeCycleCell;
  intEX: TypeCycleCell;
  MEM: TypeCycleCell;
  WB: TypeCycleCell;

  faddEX: TypeCycleCellUnit[];
  fmultEX: TypeCycleCellUnit[];
  fdivEX: TypeCycleCellUnit[];
  arrows: TypeArrowCycle[];
};

export type TypePipelineInstructions = {
  IF: { text: string; draw: boolean | TypeStall; };
  ID: { text: string; draw: boolean | TypeStall; };
  intEX: { text: string; draw: boolean | TypeStall; };
  MEM: { text: string; draw: boolean | TypeStall; };
  WB: { text: string; draw: boolean | TypeStall; };
  faddEX: { unit: number; text: string; draw: boolean | TypeStall; }[]
  fmultEX: { unit: number; text: string; draw: boolean | TypeStall; }[]
  fdivEX: { unit: number; text: string; draw: boolean | TypeStall; }[]
};

export type TypeErrorInCode = {
  line: number;
  message: string;
  severity: EnumSeverity;
}

export type TypeSimulationStep = {
  isComplete?: boolean;
  isBreakpoint?: boolean;
  step: number;
  line: number;
  isNewInstruction: boolean;
  // stage: TypeStage;

  pipeline: TypePipeline;
  registers: TypeRegisterToUpdate[];
  memory: TypeMemoryToUpdate[];
  statistics: Partial<TypeDataStatistics>;
};

export type TypeSimulationInitResponse = {
  filename: string;
  id: string;
  date: string;
  lines: number;
  canSimulate: boolean;
  errors: TypeErrorInCode[];

  machineDirectives: TypeDirectiveData[];
  machineInstructions: TypeInstructionsData[];
  runner: TypeSimulationStep[];
};

export type TypeSimulationInitRequest = {
  id: string;
  filename: string;
  date: string;
  content: string;

  breakpoints: TypeLine[];
  registers: TypeRegisterToUpdate[];
  memory: TypeMemoryToUpdate[];
};

export type TypeDataStatistics = {
  TOTAL: {
    CYCLES_EXECUTED: { cycles: number };
    ID_EXECUTED: { instructions: number };
    INSTRUCTIONS_IN_PIPELINE: { instructions_in_pipeline: number };
  };
  HARDWARE: {
    MEMORY_SIZE: { size: number };
    FADD_EX_STAGES: { num: number; cycles: number };
    FMULT_EX_STAGES: { num: number; cycles: number };
    FDIV_EX_STAGES: { num: number; cycles: number };
    FORWARDING: { enabled: boolean };
  };
  STALLS: {
    RAW_STALLS: { num: number; per: number };
    LD_STALLS: { num: number; per: number };
    BRANCH_STALLS: { num: number; per: number };
    FLOATING_POINT_STALLS: { num: number; per: number };
    WAW_STALLS: { num: number; per: number };
    STRUCTURAL_STALLS: { num: number; per: number };
    CONTROL_STALLS: { num: number; per: number };
    TRAP_STALLS: { num: number; per: number };
    TOTAL: { num: number; per: number };
  };
  CONDITIONAL: {
    TOTAL: { num: number; per: number };
    TAKEN: { num: number; per: number };
    NOT_TAKEN: { num: number; per: number };
  };
  LOAD_STORE: {
    TOTAL: { num: number; per: number };
    LOADS: { num: number; per: number };
    STORES: { num: number; per: number };
  };
  FLOATING: {
    TOTAL: { num: number; per: number };
    ADDITIONS: { num: number; per: number };
    MULTIPLICATIONS: { num: number; per: number };
    DIVISIONS: { num: number; per: number };
  };
  TRAPS: {
    TOTAL: { num: number; per: number };
  }
};
